### PR TITLE
Fix max iteration in Growing Spheres

### PR DESCRIPTION
**Pull request by kmariuszk**

### CODE REVIEW

File: src/counterfactuals/core_struct.jl [hunk 1]
```diff
@@ -67,6 +67,7 @@ function CounterfactualExplanation(
         :invalidation_rate,
         :early_stopping,
     ]
+    @assert predict_label(M, data, x) != target "The factual `x` is already in the target class. No counterfactuals can be generated."
```
Comment (kmariuszk) [inline src/counterfactuals/core_struct.jl:70]: I'm actually a bit surprised it wasn't there in the first place, I added it to simplify the growing spheres generator algorithm. Does that make sense to you @pat-alt?

File: src/generators/non_gradient_based/growing_spheres/growing_spheres.jl [hunk 2]
```diff
@@ -38,70 +38,45 @@ function growing_spheres_generation!(ce::AbstractCounterfactualExplanation)
...
     )
 
     # Repeat until there's no counterfactual points (process of removing all counterfactuals by reducing the search space)
-    while (!isnothing(counterfactual))
+    while (!isnothing(counterfactual) && ce.search[:iteration_count] < ce.convergence[:max_iter])
         η /= 2
         a₀ = convert(eltype(factual), 0.0)
 
         counterfactual_candidates = hyper_sphere_coordinates(n, factual, a₀, η)
+        update_path(ce, counterfactual_candidates)
```
Comment (kmariuszk) [inline src/generators/non_gradient_based/growing_spheres/growing_spheres.jl:61]: Now we're updating path every time we consider a new candidate for a counterfactual

File: src/generate_counterfactual/growing_spheres.jl [hunk 3]
```diff
@@ -19,5 +19,8 @@ function generate_counterfactual(
     # growing spheres does not support encodings, thus x′ is just s′
     ce.x′ = ce.s′
 
+    ce.search[:converged] = converged(ce)
+    ce.search[:terminated] = terminated(ce)
+
```
Comment (VincentPikand) [inline src/generate_counterfactual/growing_spheres.jl:24]: resolved by #360

File: src/counterfactuals/convergence.jl [hunk 4]
```diff
@@ -26,8 +26,10 @@ A convenience method to determine if the counterfactual search has converged. Th
 """
 
 function converged(ce::CounterfactualExplanation)
-    if ce.generator isa GrowingSpheresGenerator || ce.generator isa FeatureTweakGenerator
+    if ce.generator isa FeatureTweakGenerator
         conv = ce.search[:converged]
+    elseif ce.generator isa GrowingSpheresGenerator
+        conv = !isnothing(ce.s′)
```
Comment (VincentPikand) [inline src/counterfactuals/convergence.jl:32]: resolved by #360

**Review by VincentPikand — COMMENTED** [review]
This should be put on pause until rauno has merged #360, which completely refactors convergence.

File: src/generators/non_gradient_based/growing_spheres/growing_spheres.jl [hunk 5]
```diff
@@ -38,73 +38,56 @@ function growing_spheres_generation!(ce::AbstractCounterfactualExplanation)
...
-        if max_iter == 0
-            break
-        end
     end
 
-    # Update path
-    ce.search[:iteration_count] += n
-    for i in eachindex(counterfactual_candidates[1, :])
-        push!(ce.search[:path], reshape(counterfactual_candidates[:, i], :, 1))
+    if !isnothing(counterfactual)
```
[the remark this answers is not in the document]
Comment (kmariuszk) [reply src/generators/non_gradient_based/growing_spheres/growing_spheres.jl:87]: It is not due to the second condition `&& ce.search[:iteration_count] < ce.convergence[:max_iter]`

**Comment by kmariuszk** [comment]
> This should be put on pause until rauno has merged https://github.com/JuliaTrustworthyAI/CounterfactualExplanations.jl/pull/360, which completely refactors convergence.

Cool, hope Rauno can find something useful in my PR. Let's leave it open till #360 is merged and start from there. I'll make it a draft for now.

**Comment by VincentPikand** [comment]
Additionally, I am now sold on dispatching on `generate_perturbations` (or `update!`, we can discuss this if necessary). However, imo the crux is how to we make it clear to developers and users that things like `apply_mutability`, `apply_domain_constraints!`, etc are only applied when they are "enabled" or otherwise specified. At first glance, it might be confusing why do we call functions like that if they were not requested.


```julia
function update!(ce::CounterfactualExplanation)

    # Generate peturbations:
    Δs′ = Generators.generate_perturbations(ce.generator, ce)
    Δs′ = apply_mutability(ce, Δs′)         # mutability constraints
    s′ = ce.s′ + Δs′                        # new proposed state

    # Updates:
    ce.s′ = s′                                                  # update counterfactual
    ce.x′ = decode_state(ce)                                    # decoded counterfactual state
    apply_domain_constraints!(ce)                               # apply domain constraints
    _times_changed = reshape(
        decode_state(ce, Δs′) .!= 0, size(ce.search[:times_changed_features])
    )
    ce.search[:times_changed_features] += _times_changed        # update number of times feature has been changed
    ce.search[:mutability] = Generators.mutability_constraints(ce.generator, ce)
    ce.search[:iteration_count] += 1                            # update iteration counter   
    ce.search[:path] = [ce.search[:path]..., ce.s′]
    ce.search[:converged] = converged(ce)
    return ce.search[:terminated] = terminated(ce)
end
```

This means that growing spheres needs a bit of a refactoring as well to fit into that framework, so I suggest we make this PR about that.

@pat-alt @kmariuszk @kmariuszk 
Like if you approve.

**Comment by kmariuszk** [comment]
> This means that growing spheres needs a bit of a refactoring as well to fit into that framework, so I suggest we make this PR about that.

The idea sounds interesting to me, I'm also leaning towards this solution, I'm pretty sure we can find a way to make it more obvious for the end users. Although, should I start working on that before Rauno's PR is merged?

**Comment by VincentPikand** [comment]
> > This means that growing spheres needs a bit of a refactoring as well to fit into that framework, so I suggest we make this PR about that.
> 
> The idea sounds interesting to me, I'm also leaning towards this solution, I'm pretty sure we can find a way to make it more obvious for the end users. Although, should I start working on that before Rauno's PR is merged?

It should be merged pretty soon, give it a moment

**Pull request closed without merging**